src/main/flight/imu.c

```diff
@@ -407,6 +394,73 @@ static bool imuIsAccelerometerHealthy(void)
     return (81 < accMagnitude) && (accMagnitude < 121);
 }
 
+// Calculate the dcmKpGain to use. When armed, the gain is imuRuntimeConfig.dcm_kp * 1.0 scaling.
+// When disarmed after initial boot, the scaling is set to 10.0 for the first 20 seconds to speed up initial convergence.
+// After disarming we want to quickly reestablish convergence to deal with the attitude estimation being incorrect due to a crash.
+//   - wait for a 250ms period of low gyro activity to ensure the craft is not moving
+//   - use a large dcmKpGain value for 500ms to allow the attitude estimate to quickly converge
+//   - reset the gain back to the standard setting
+float imuCalcKpGain(timeUs_t currentTimeUs, bool useAcc, float gyroX, float gyroY, float gyroZ)
+{
+    static bool lastArmState = false;
+    static timeUs_t gyroQuietPeriodTimeEnd = 0;
+    static timeUs_t attitudeResetTimeEnd = 0;
+    static bool attitudeResetCompleted = false;
+    float ret;
+    bool attitudeResetActive = false;
+
+    const bool armState = ARMING_FLAG(ARMED);
+
+    if (!armState) {
+        if (!STATE(FIXED_WING)) {  // disable for fixed wing. probably wouldn't hurt anything but not needed or tested
+            if (lastArmState) {   // Just disarmed; start the gyro quiet period
+                gyroQuietPeriodTimeEnd = currentTimeUs + ATTITUDE_RESET_QUIET_TIME;
+                attitudeResetTimeEnd = 0;
+                attitudeResetCompleted = false;
+            }
+
+            // If gyro activity exceeds the threshold then restart the quiet period.
+            // Also, if the attitude reset has been complete and there is subsequent gyro activity then
+            // start the reset cycle again. This addresses the case where the pilot rights the craft after a crash.
+            if ((attitudeResetTimeEnd > 0) || (gyroQuietPeriodTimeEnd > 0) || attitudeResetCompleted) {
+                if ((fabsf(gyroX) > ATTITUDE_RESET_GYRO_LIMIT)
```

Comment:
(using L1 would work here - fabsf(gyroX) + fabsf(gyroY) + fabsf(gyroZ). Or compute true magSquared)

Reply:
The logic is looking for any axis to be above the threshold - not the sum of multiple axes.  Just seemed simpler and was easier to debug in blackbox logs by looking at the gyro traces.  I suppose I could calculate the spin rate `sqrtf(sq(gx) + sq(gy) + sq(gz))` but that would seem more expensive than the condition.  Also I'd have to come up with an appropriate rate threshold.

Reply:
Yes, it's not important... L1 would be a bit cheaper than  L2/euclidean and close enough. And it would be actually few ticks faster than 3x comparison

(BTW: no need for sqrt, `sq(gx) + sq(gy) + sq(gz) > sq(ATTITUDE_RESET_GYRO_LIMIT)` )

Reply:
I understand.  But since this is all running under the disarmed state where performance is not ultra-critical I think I'd prefer to leave the conditional for ease of understanding.